### PR TITLE
Improve some typing in test_runner

### PR DESCRIPTION
**Pull request by tristan957**
Fixes some types, and replaces deprecated types.

### CODE REVIEW

**Comment by tristan957** [comment]
I find pyright to be a lot more useful than mypy.

File: test_runner/fixtures/benchmark_fixture.py [hunk 1]
```diff
@@ -23,6 +23,10 @@
 from fixtures.log_helper import log
 from fixtures.neon_fixtures import NeonPageserver
 
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+    from typing import Callable, ClassVar, Optional
```
Comment (hlinnaka) [inline test_runner/fixtures/benchmark_fixture.py:28]: Is there some harm in always importing these? I.e. why do these need to be guarded by TYPE_CHECKING?

(comment applies to all such blocks in the PR)

Comment (tristan957) [reply test_runner/fixtures/benchmark_fixture.py:28]: It's a micro-optimization to help tests at runtime. The Python parser will skip `TYPE_CHECKING` blocks.

**Comment by tristan957** [comment]
The fact that mypy passes on my machine, but not CI is very annoying :'(.

File: test_runner/fixtures/benchmark_fixture.py (hunk 1 above)
Comment (tristan957) [reply test_runner/fixtures/benchmark_fixture.py:28]: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Comment (tristan957) [reply test_runner/fixtures/benchmark_fixture.py]: We could move a lot more, but it requires more developer discipline. I just got some low hanging fruit in relation to this PR. I wonder if there are tools to detect imports that could move to `TYPE_CHECKING` blocks.

**Comment by tristan957** [comment]
Almost all the regression test file changes you see in this PR are a result of me being blind and not reading the logs correctly. I think I'll probably reformat the commit message at the time of merge to say something a little more generic.

**Review by bayandin — COMMENTED** [review]
Can we enforce using the same style for type annotation everywhere with `ruff` / `mypy`? (for example using `dict` instead of `Dict`)

Otherwise, I'm afraid, the codebase will be more inconsistent.

**Comment by tristan957** [comment]
> Can we enforce using the same style for type annotation everywhere with `ruff` / `mypy`? (for example using `dict` instead of `Dict`)
> 
> Otherwise, I'm afraid, the codebase will be more inconsistent.

https://docs.astral.sh/ruff/rules/deprecated-import/. I detect these issues with pyright. I would assume that mypy as this support?

I can open an additional PR to fix it everywhere, or do it as a precursor PR. What would you prefer?

**Comment by bayandin** [comment]
> I can open an additional PR to fix it everywhere, or do it as a precursor PR. What would you prefer?

It would be nice to open another PR with enabled checks and applying fixes to make sure we keep this code consistent.
Thank you!

**Review by hlinnaka — APPROVED** [review]

**Review by bayandin — COMMENTED** [review]
Are there any linters to ensure consistensy for the future?
For example. in the PR we introduce and remove `Optional` in different places:

```diff
-    def hash_extracted(reader: Union[IO[bytes], None]) -> bytes:
+    def hash_extracted(reader: Optional[IO[bytes]]) -> bytes:
```
And at the same point

```diff
-    def __init__(self, neon_api: NeonAPI, pg_version: PgVersion, project_id: Optional[str]):
+    def __init__(self, neon_api: NeonAPI, pg_version: PgVersion, project_id: str | None):
```

**Comment by tristan957** [comment]
See https://github.com/neondatabase/neon/pull/9349

**Pull request merged**